include/cantera/kinetics/MultiRate.h

```diff
@@ -51,73 +45,68 @@ class MultiRateBase
     //! @param bulk  object representing bulk phase
     //! @param kf  array of rate constants
     virtual void getRateConstants(const ThermoPhase& bulk,
-                                  double* kf) const = 0;
+                                  double* kf, double* concm) const = 0;
 
     //! Update data common to reaction rates of a specific type
     //! @param bulk  object representing bulk phase
-    virtual void update(const ThermoPhase& bulk) = 0;
+    virtual void update(const ThermoPhase& bulk, double* concm) = 0;
 };
 
 
 //! A class template handling all reaction rates specific to `BulkKinetics`.
-/**
- * @warning This class is an experimental part of the %Cantera API and
- *     may be changed or removed without notice.
- */
 template <class RateType, class DataType>
 class MultiBulkRates final : public MultiRateBase
 {
 public:
     virtual void add(const size_t rxn_index,
                      ReactionRateBase& rate) override {
-        m_indices[rxn_index] = m_rates.size();
-        m_rates.push_back(dynamic_cast<RateType&>(rate));
-        m_rxn.push_back(rxn_index);
+        m_indices[rxn_index] = m_rxn_rates.size();
+        m_rxn_rates.emplace_back(rxn_index, static_cast<RateType&>(rate));
```

Comment:
The `dynamic_cast` here was correct, because it can't be guaranteed that when this method is called `rate` is actually of the type `RateType` and not some other type derived from `ReactionRateBase`. I don't think that this particular code is so performance sensitive as to make it worth foregoing the runtime type checking.